File: packages/react-native/android/src/main/java/com/margelo/nitro/blasphem/HybridBlasphemAssets.kt
package com.margelo.nitro.blasphem

import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise

/**
 * Reads `manifest.json`, `<code>.pack`, and `<code>.detect` from the app's
 * assets under `assets/blasphem/`.
 */
@DoNotStrip
class HybridBlasphemAssets : HybridBlasphemAssetsSpec() {
  override fun readBundled(name: String): Promise<ArrayBuffer> {
    return Promise.parallel {
      require(name == java.io.File(name).name) { "BLASPHEM_FETCH_FAILED: Invalid bundled filename" }
      val bytes = BlasphemFileIO.context().assets.open("blasphem/$name").use { it.readBytes() }
      buffer(bytes)
    }
  }

  override fun readManifest(url: String, refresh: Boolean): Promise<ArrayBuffer> =
    Promise.parallel { buffer(BlasphemDownloadStore.manifest(url, refresh)) }

  override fun commitManifest(url: String, bytes: ArrayBuffer): Promise<Unit> {
    val source = bytes.getBuffer(true).duplicate()
    val copy = ByteArray(source.remaining())
    source.get(copy)
    return Promise.parallel { BlasphemDownloadStore.commit(url, copy) }
  }

  override fun readDownloaded(url: String, expected: DownloadIntegrity): Promise<ArrayBuffer> =
    Promise.parallel { buffer(BlasphemDownloadStore.downloaded(url, expected)) }

  private fun buffer(bytes: ByteArray): ArrayBuffer =
    ArrayBuffer.allocate(bytes.size).also { it.getBuffer(false).put(bytes) }
}

File: packages/react-native/ios/HybridBlasphemAssets.swift
import Foundation
import NitroModules

/// Reads `manifest.json`, `<code>.pack`, and `<code>.detect` from the app
/// bundle. Add the files to the target as a folder reference named `blasphem`,
/// or at the bundle root.
class HybridBlasphemAssets: HybridBlasphemAssetsSpec {
  func readBundled(name: String) throws -> Promise<ArrayBuffer> {
    return Promise.async {
      let url = Bundle.main.url(forResource: name, withExtension: nil, subdirectory: "blasphem")
        ?? Bundle.main.url(forResource: name, withExtension: nil)
      guard let url else {
        throw RuntimeError.error(withMessage: "BLASPHEM_FETCH_FAILED: \(name) is not in the app bundle")
      }
      let data = try Data(contentsOf: url)
      return try ArrayBuffer.copy(data: data)
    }
  }
}
